### PR TITLE
Upgrade react-native 0.72.0-rc.5

### PR DESCRIPTION
**Pull request by Kudo**
# Why

upgrade react-native 0.72 for sdk 49
close ENG-8011

# How

- bump package versions
  - `react-native 0.71.3 -> 0.72.0-rc.5`
  - `metro-react-native-babel-preset 0.73.9 -> 0.76.5`
- [bare-expo][templates][fabric-tester] migrate template base on [upgrade-helper](https://react-native-community.github.io/upgrade-helper/?from=0.71.7&to=0.72.0-rc.5)
- [expo-template-tabs] remove the metro version overrides for expo-router.
- [core][dev-laucher][dev-menu][media-library][screen-orientation][splash-screen][updates-interface][updates] add the `install_modules_dependencies` to support new architecture + use_frameworks!
- [core][autolinking] fix some new architecture error on ios
- [react-native-lab] update our fork to 0.72.0-rc.5 based
- [go][tools] fix **react-native-lab/react-native/packages/react-native** path move because of react-native's repo monorepo changes
- [go][android] fix gradle 8 errors
- [go][ios] add `RCT_REMOTE_PROFILE=0` to fix the `RCT_ENABLE_INSPECTOR needs to be set to fulfill RCT_REMOTE_PROFILE` build error
- [ncl] remove `ProgressViewIOS` / `ProgressBarAndroid` since they are deprecated/removed in 0.72
- [dev-menu][dev-launcher] rebuild bundles

# Note

- react-native-web is not bumped because of the [issue](https://github.com/necolas/react-native-web/issues/2523), so it's still react-native-web@~0.18.10.
- currently disable ci typecheck for @expo/cli because of upstream metro typescript support. i'll have another pr to fix those errors.
- updates e2e ci on android is broken at [here](https://github.com/expo/expo/blob/fada3d764957779fbfc3d7b723d185db1d933d95/packages/expo-updates/e2e/fixtures/Updates.e2e.ts#L518). i doubt if that's related to the react scheduler change. i'd disabled the failed test case.
- the react-native upstream [migrated away the `@types/jest`](https://github.com/facebook/react-native/pull/36068). i was afraiding that will be a breaking change to the existing jest test code since it requires the explicit `@jest/globals` import. i didn't do this in this upgrade.

# Test Plan

- ✅ fabric-tester (without expo-dev-client)
- ✅ ci passed. there are some errors which are known:
  - updates e2e on android: as mentioned above
  - ios expo go on eas build: versioned expo go are broken on eas build m1 worker. this is also happening on main.
  - android client: no space left on the ubuntu worker. this is also happening on main.
- ✅ bare-expo
- ✅ unversioned expo go + ncl

# Checklist

- [x] Documentation is up to date to reflect these changes (eg: https://docs.expo.dev and README.md).
- [x] Conforms with the [Documentation Writing Style Guide](https://github.com/expo/expo/blob/main/guides/Expo%20Documentation%20Writing%20Style%20Guide.md)
- [x] This diff will work correctly for `npx expo prebuild` & EAS Build (eg: updated a module plugin).

### CODE REVIEW

File: .github/workflows/android-instrumentation-tests.yml [hunk 1]
```diff
@@ -35,7 +35,7 @@ concurrency:
 jobs:
   test:
     runs-on: macos-12
-    timeout-minutes: 120
+    timeout-minutes: 150
```
Comment (tsapeta) [inline .github/workflows/android-instrumentation-tests.yml:38]: 📈

File: .gitignore [hunk 2]
```diff
@@ -65,6 +65,7 @@ xcuserdata
 ReactAndroid-temp.aar
 /android/prebuiltHermes
 /android/versioned-abis/**/*.aar
+/android/packages/react-native/sdks/hermes
```
Comment (tsapeta) [inline .gitignore:68]: Do we have such folder in the root? 🤔 Isn't it only in react-native-lab?

File: packages/expo-dev-launcher/expo-dev-launcher.podspec [hunk 3]
```diff
@@ -61,6 +61,8 @@ Pod::Spec.new do |s|
   s.dependency "expo-dev-menu"
   s.dependency "ExpoModulesCore"
 
+  install_modules_dependencies(s)
```
[the remark this answers is not in the document]
Comment (tsapeta) [reply packages/expo-dev-launcher/expo-dev-launcher.podspec:64]: Just leaving a note that this check is required for `pod ipc spec` command that converts the podspec into a JSON. We use it in versioning, but it runs without executing the Podfile and in that context React Native script that defines this function is not imported.

File: packages/expo-image/CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,7 @@
 
 - Add prefetch implementation on web. ([#22630](https://github.com/expo/expo/pull/22630) by [@aleqsio](https://github.com/aleqsio))
 - Add `ImageBackground` component. ([#22347](https://github.com/expo/expo/pull/22347) by [@alanjhughes](https://github.com/alanjhughes))
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-image/CHANGELOG.md:11]: For SDK packages I would rather say this 👇 
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo-media-library/CHANGELOG.md [hunk 5]
```diff
@@ -10,6 +10,8 @@
 
 ### 🎉 New features
 
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-media-library/CHANGELOG.md:13]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo-modules-autolinking/CHANGELOG.md [hunk 6]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-modules-autolinking/CHANGELOG.md:9]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo-modules-core/CHANGELOG.md [hunk 7]
```diff
@@ -12,6 +12,7 @@
 
 - Added `ReactActivityHandler.getDelayLoadAppHandler` interface on Android. ([#20273](https://github.com/expo/expo/pull/20273) by [@kudo](https://github.com/kudo))
 - [iOS] Introduced shared refs – a way to pass native objects among different independent modules. ([#22583](https://github.com/expo/expo/pull/22583) by [@tsapeta](https://github.com/tsapeta))
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-modules-core/CHANGELOG.md:15]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo-screen-orientation/CHANGELOG.md [hunk 8]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-screen-orientation/CHANGELOG.md:9]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo-splash-screen/CHANGELOG.md [hunk 9]
```diff
@@ -6,6 +6,8 @@
 
 ### 🎉 New features
 
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-splash-screen/CHANGELOG.md:9]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo-updates/CHANGELOG.md [hunk 10]
```diff
@@ -11,6 +11,7 @@
 ### 🎉 New features
 
 - [Android] Load updates in background thread and prevent ANR from initial launch. ([#20273](https://github.com/expo/expo/pull/20273) by [@kudo](https://github.com/kudo))
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo-updates/CHANGELOG.md:14]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

File: packages/expo/CHANGELOG.md [hunk 11]
```diff
@@ -11,6 +11,7 @@
 ### 🎉 New features
 
 - Added `ReactActivityHandler.getDelayLoadAppHandler` interface on Android. ([#20273](https://github.com/expo/expo/pull/20273) by [@kudo](https://github.com/kudo))
+- Update `react-native` to 0.72.0. ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```
Comment (tsapeta) [inline packages/expo/CHANGELOG.md:14]:
```suggestion
- Added support for React Native 0.72 ([#22588](https://github.com/expo/expo/pull/22588) by [@kudo](https://github.com/kudo))
```

**Review by tsapeta — APPROVED** [review]

File: .github/workflows/android-instrumentation-tests.yml (hunk 1 above)
Comment (Kudo) [reply .github/workflows/android-instrumentation-tests.yml:38]: building hermes requires more time than usual 🤦‍♂️

**Review by douglowder — APPROVED** [review]

**Pull request merged**